python/cudf/cudf/core/column/column.py

```diff
@@ -1948,8 +1948,23 @@ def as_column(
                 raise TypeError(
                     f"Cannot convert a {inferred_dtype} of object type"
                 )
-            elif nan_as_null is False and (
-                pd.isna(arbitrary).any()
+            elif inferred_dtype == "boolean":
+                if cudf.get_option("mode.pandas_compatible"):
+                    if not (
+                        dtype is not None
+                        and dtype == np.dtype("bool")
+                        and not pd.isna(arbitrary).any()
+                    ):
+                        raise MixedTypeError(
+                            f"Cannot have mixed values with {inferred_dtype}"
+                        )
+                elif nan_as_null is False and _has_any_nan(arbitrary):
+                    raise MixedTypeError(
+                        f"Cannot have mixed values with {inferred_dtype}"
+                    )
+            elif (
+                nan_as_null is False
```

Comment:
Don't compare to booleans with `is`.
```suggestion
                not nan_as_null
```

Reply:
Same as https://github.com/rapidsai/cudf/pull/15709/files#r1600749515